### PR TITLE
Fix golang + linux + fstat

### LINKED ISSUE #2254: go application mirroring failed with error
**Bug Description**

When I try to start mirrord with go application, I got next error:
```
thread '<unnamed>' panicked at mirrord/layer/src/detour.rs:98:22:
called `Option::unwrap()` on a `None` value
```
**Environment:**
  local:
  Description:	Debian GNU/Linux 11 (bullseye)
  GoLand 2023.3.4 (Build #GO-233.14475.38, built on February 14, 2024)+plugin mirrord 3.58.0
  go version go1.19 linux/amd64
  
  remote:
  K8s Rev: v1.27.9-gke.1092000

**Config:**
```
{
  "kube_context": "{my_context}",
  "agent": {
    "check_out_of_pods": false,
    "namespace": "{my_namespace}",
    "log_level": "mirrord=debug,warn"
  },
  "target": {
    "namespace": "{my_namespace}",
    "path": {
      "pod": "{my_pod_name}",
      "container": "{my_container_name}"
    }
  },
  "feature": {
    "network": {
      "incoming": {
        "mode": "steal",
        "http_filter": {
          "header_filter": "Username: username"
        }
      },
      "outgoing": {
        "ignore_localhost": true
      }
    },
    "fs": "localwithoverrides"
  }
}
```
**Logs:**
```
GOROOT=/usr/local/src/go/1.19/go #gosetup
GOPATH=/home/useername/go #gosetup
/usr/local/src/go/1.19/go/bin/go build -o /home/useername/.cache/JetBrains/GoLand2023.3/tmp/GoLand/app -gcflags all=-N -l git/service/app #gosetup
/home/useername/.local/share/JetBrains/Toolbox/apps/Goland/ch-0/233.14015.113/plugins/go-plugin/lib/dlv/linux/dlv --listen=127.0.0.1:43375 --headless=true --api-version=2 --check-go-version=false --only-same-user=false exec /home/useername/.cache/JetBrains/GoLand2023.3/tmp/GoLand/app --
API server listening at: 127.0.0.1:43375
thread '<unnamed>' panicked at mirrord/layer/src/detour.rs:98:22:
called `Option::unwrap()` on a `None` value
stack backtrace:
   0:     0x7f43214ef17c - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hca9d80fb6e5ad5fe
   1:     0x7f432151ae20 - core::fmt::write::h7acb2cb5522bc491
   2:     0x7f43214ec69f - std::io::Write::write_fmt::h19435141596ff66c
   3:     0x7f43214eef64 - std::sys_common::backtrace::print::h120ea2df27b2be8f
   4:     0x7f43214f0437 - std::panicking::default_hook::{{closure}}::hd7e3d5fa26e2ca93
   5:     0x7f43214f019f - std::panicking::default_hook::h541732093fb638b5
   6:     0x7f43214f08b8 - std::panicking::rust_panic_with_hook::h10b8d62d65c23db7
   7:     0x7f43214f0769 - std::panicking::begin_panic_handler::{{closure}}::h2bc5b33298108b9d
   8:     0x7f43214ef646 - std::sys_common::backtrace::__rust_end_short_backtrace::h4d52701c18fca37f
   9:     0x7f43214f0502 - rust_begin_unwind
  10:     0x7f4320a006e5 - core::panicking::panic_fmt::hc7d083b823bc25c6
  11:     0x7f4320a007a3 - core::panicking::panic::h54d458dc4fbcf3ea
  12:     0x7f4320a88132 - core::option::Option<T>::unwrap::ha3b16f2c1e38c70e
  13:     0x7f4320ba6865 - <mirrord_layer::detour::HookFn<T> as core::ops::deref::Deref>::deref::h0c130757e3df039a
  14:     0x7f4320a59cdf - mirrord_layer::file::hooks::fstat_detour::{{closure}}::hd0d0ffd5cb8f6967
  15:     0x7f4320bb28d2 - mirrord_layer::detour::Detour<S>::unwrap_or_bypass_with::hf8525d06d01af018
  16:     0x7f4320a59c48 - mirrord_layer::file::hooks::fstat_detour::hf910afc17afa42e0
  17:     0x7f4320ace350 - c_abi_syscall6_handler
  18:     0x7f4320b5da8b - mirrord_layer::go::go_hooks::go_syscall_new_detour::hd7d0061902bccdce
  19:                0x0 - <unknown>
fatal runtime error: failed to initiate panic, error 5
SIGABRT: abort
PC=0x7f4320664ce1 m=0 sigcode=18446744073709551610

goroutine 0 [idle]:
runtime: g 0: unknown pc 0x7f4320664ce1
stack: frame={sp:0x7ffdbf324320, fp:0x0} stack=[0x7ffdbeb26150,0x7ffdbf325160)
0x00007ffdbf324220:  0x0000000000000000  0x0000000000000000 
0x00007ffdbf324230:  0x0000000000000000  0x0000000000000000 
0x00007ffdbf324240:  0x0000000000000000  0x0000000000000000 
0x00007ffdbf324250:  0x0000000000000000  0x0000000000000000 
0x00007ffdbf324260:  0x0000000000000000  0x0000000000000000 
0x00007ffdbf324270:  0x0000000000000000  0x0000000000000000 
0x00007ffdbf324280:  0x00007ffdbf3250d8  0x0000000000000000 
0x00007ffdbf324290:  0x00007ffdbf3250e0  0x00007f4320b5da8b 
0x00007ffdbf3242a0:  0x0000000000000000  0x0000000000000000 
0x00007ffdbf3242b0:  0x0000000000000000  0x00007f4320b5da00 
0x00007ffdbf3242c0:  0x4000000000000000  0x0000000000000000 
0x00007ffdbf3242d0:  0x0000000000000000  0x0000000000000000 
0x00007ffdbf3242e0:  0x0000000000000000  0x0000000000000000 
0x00007ffdbf3242f0:  0x00007ffdbf3250c8  0x00007ffdbf3250d0 
0x00007ffdbf324300:  0x0000000000fc659a <github.com/uber/jaeger-client-go.(*Tracer).startSpanInternal+0x000000000000039a>  0x00007f42f85f5010 
0x00007ffdbf324310:  0x000000c00013f140  0x0000000000000004 
0x00007ffdbf324320: <0x0000000000000000  0x0000000000000007 
0x00007ffdbf324330:  0x0000000000000000  0x000000c000209230 
0x00007ffdbf324340:  0x0000000000000002  0x0000000000000002 
0x00007ffdbf324350:  0x0000000000000004  0x0000000000000008 
0x00007ffdbf324360:  0x0000000000000000  0x000000c00013f164 
0x00007ffdbf324370:  0x0000000000000000  0x0000000000000000 
0x00007ffdbf324380:  0x0000000000000000  0x0000000000000000 
0x00007ffdbf324390:  0x0000000000000000  0x00007f4320b2cb0e 
0x00007ffdbf3243a0:  0xfffffffe7fffffff  0xffffffffffffffff 
0x00007ffdbf3243b0:  0xffffffffffffffff  0xffffffffffffffff 
0x00007ffdbf3243c0:  0xffffffffffffffff  0xffffffffffffffff 
0x00007ffdbf3243d0:  0xffffffffffffffff  0xffffffffffffffff 
0x00007ffdbf3243e0:  0xffffffffffffffff  0xffffffffffffffff 
0x00007ffdbf3243f0:  0xffffffffffffffff  0xffffffffffffffff 
0x00007ffdbf324400:  0xffffffffffffffff  0xffffffffffffffff 
0x00007ffdbf324410:  0xffffffffffffffff  0xffffffffffffffff 
runtime: g 0: unknown pc 0x7f4320664ce1
stack: frame={sp:0x7ffdbf324320, fp:0x0} stack=[0x7ffdbeb26150,0x7ffdbf325160)
0x00007ffdbf324220:  0x0000000000000000  0x0000000000000000 
0x00007ffdbf324230:  0x0000000000000000  0x0000000000000000 
0x00007ffdbf324240:  0x0000000000000000  0x0000000000000000 
0x00007ffdbf324250:  0x0000000000000000  0x0000000000000000 
0x00007ffdbf324260:  0x0000000000000000  0x0000000000000000 
0x00007ffdbf324270:  0x0000000000000000  0x0000000000000000 
0x00007ffdbf324280:  0x00007ffdbf3250d8  0x0000000000000000 
0x00007ffdbf324290:  0x00007ffdbf3250e0  0x00007f4320b5da8b 
0x00007ffdbf3242a0:  0x0000000000000000  0x0000000000000000 
0x00007ffdbf3242b0:  0x0000000000000000  0x00007f4320b5da00 
0x00007ffdbf3242c0:  0x4000000000000000  0x0000000000000000 
0x00007ffdbf3242d0:  0x0000000000000000  0x0000000000000000 
0x00007ffdbf3242e0:  0x0000000000000000  0x0000000000000000 
0x00007ffdbf3242f0:  0x00007ffdbf3250c8  0x00007ffdbf3250d0 
0x00007ffdbf324300:  0x0000000000fc659a <github.com/uber/jaeger-client-go.(*Tracer).startSpanInternal+0x000000000000039a>  0x00007f42f85f5010 
0x00007ffdbf324310:  0x000000c00013f140  0x0000000000000004 
0x00007ffdbf324320: <0x0000000000000000  0x0000000000000007 
0x00007ffdbf324330:  0x0000000000000000  0x000000c000209230 
0x00007ffdbf324340:  0x0000000000000002  0x0000000000000002 
0x00007ffdbf324350:  0x0000000000000004  0x0000000000000008 
0x00007ffdbf324360:  0x0000000000000000  0x000000c00013f164 
0x00007ffdbf324370:  0x0000000000000000  0x0000000000000000 
0x00007ffdbf324380:  0x0000000000000000  0x0000000000000000 
0x00007ffdbf324390:  0x0000000000000000  0x00007f4320b2cb0e 
0x00007ffdbf3243a0:  0xfffffffe7fffffff  0xffffffffffffffff 
0x00007ffdbf3243b0:  0xffffffffffffffff  0xffffffffffffffff 
0x00007ffdbf3243c0:  0xffffffffffffffff  0xffffffffffffffff 
0x00007ffdbf3243d0:  0xffffffffffffffff  0xffffffffffffffff 
0x00007ffdbf3243e0:  0xffffffffffffffff  0xffffffffffffffff 
0x00007ffdbf3243f0:  0xffffffffffffffff  0xffffffffffffffff 
0x00007ffdbf324400:  0xffffffffffffffff  0xffffffffffffffff 
0x00007ffdbf324410:  0xffffffffffffffff  0xffffffffffffffff 

goroutine 1 [syscall, locked to thread]:
syscall.Syscall(0x5, 0x8, 0xc000177078, 0x0)
        /usr/local/src/go/1.19/go/src/syscall/syscall_linux.go:68 +0x4b fp=0xc0001cd398 sp=0xc0001cd320 pc=0x4b2d2b
syscall.Fstat(0x8, 0xc000177078)
        /usr/local/src/go/1.19/go/src/syscall/zsyscall_linux_amd64.go:1092 +0x3f fp=0xc0001cd3f0 sp=0xc0001cd398 pc=0x4b115f
internal/poll.(*FD).Fstat.func1()
        /usr/local/src/go/1.19/go/src/internal/poll/fd_unix.go:687 +0x46 fp=0xc0001cd448 sp=0xc0001cd3f0 pc=0x540c06
internal/poll.ignoringEINTR(0xc0001cd520)
        /usr/local/src/go/1.19/go/src/internal/poll/fd_posix.go:74 +0x2e fp=0xc0001cd478 sp=0xc0001cd448 pc=0x53960e
internal/poll.(*FD).Fstat(0xc00013c3c0, 0xc000177078)
        /usr/local/src/go/1.19/go/src/internal/poll/fd_unix.go:686 +0x146 fp=0xc0001cd548 sp=0xc0001cd478 pc=0x540b46
os.(*File).Stat(0xc00013a2c0)
        /usr/local/src/go/1.19/go/src/os/stat_unix.go:20 +0xc9 fp=0xc0001cd5d8 sp=0xc0001cd548 pc=0x551d49
os.ReadFile({0x12e89e1, 0xf})
        /usr/local/src/go/1.19/go/src/os/file.go:680 +0x1dc fp=0xc0001cd7b8 sp=0xc0001cd5d8 pc=0x54d53c
io/ioutil.ReadFile({0x12e89e1, 0xf})
        /usr/local/src/go/1.19/go/src/io/ioutil/ioutil.go:37 +0x75 fp=0xc0001cd850 sp=0xc0001cd7b8 pc=0x664d35
github.com/rs/xid.readPlatformMachineID()
        /home/useername/go/pkg/mod/github.com/rs/xid@v1.4.0/hostid_linux.go:8 +0x5e fp=0xc0001cd8f8 sp=0xc0001cd850 pc=0xad7f7e
github.com/rs/xid.readMachineID()
        /home/useername/go/pkg/mod/github.com/rs/xid@v1.4.0/id.go:115 +0x89 fp=0xc0001cdab0 sp=0xc0001cd8f8 pc=0xad8369
github.com/rs/xid.init()
        /home/useername/go/pkg/mod/github.com/rs/xid@v1.4.0/id.go:82 +0x25 fp=0xc0001cdac0 sp=0xc0001cdab0 pc=0xadb365
runtime.doInit(0x1889020)
        /usr/local/src/go/1.19/go/src/runtime/proc.go:6321 +0x132 fp=0xc0001cdbf0 sp=0xc0001cdac0 pc=0x450132
runtime.doInit(0x188c6c0)
        /usr/local/src/go/1.19/go/src/runtime/proc.go:6298 +0x79 fp=0xc0001cdd20 sp=0xc0001cdbf0 pc=0x450079
runtime.doInit(0x18887a0)
        /usr/local/src/go/1.19/go/src/runtime/proc.go:6298 +0x79 fp=0xc0001cde50 sp=0xc0001cdd20 pc=0x450079
runtime.doInit(0x188fe60)
        /usr/local/src/go/1.19/go/src/runtime/proc.go:6298 +0x79 fp=0xc0001cdf80 sp=0xc0001cde50 pc=0x450079
runtime.main()
        /usr/local/src/go/1.19/go/src/runtime/proc.go:233 +0x199 fp=0xc0001cdfe0 sp=0xc0001cdf80 pc=0x442d19
runtime.goexit()
        /usr/local/src/go/1.19/go/src/runtime/asm_amd64.s:1594 +0x1 fp=0xc0001cdfe8 sp=0xc0001cdfe0 pc=0x4755c1

goroutine 2 [force gc (idle)]:
runtime.gopark(0x1338038, 0x18e2e70, 0x11, 0x14, 0x1)
        /usr/local/src/go/1.19/go/src/runtime/proc.go:363 +0xfd fp=0xc00008ef88 sp=0xc00008ef58 pc=0x44317d
runtime.goparkunlock(0x0?, 0x0?, 0x0?, 0x0?)
        /usr/local/src/go/1.19/go/src/runtime/proc.go:369 +0x2a fp=0xc00008efb8 sp=0xc00008ef88 pc=0x44320a
runtime.forcegchelper()
        /usr/local/src/go/1.19/go/src/runtime/proc.go:302 +0xa5 fp=0xc00008efe0 sp=0xc00008efb8 pc=0x442fa5
runtime.goexit()
        /usr/local/src/go/1.19/go/src/runtime/asm_amd64.s:1594 +0x1 fp=0xc00008efe8 sp=0xc00008efe0 pc=0x4755c1
created by runtime.init.6
        /usr/local/src/go/1.19/go/src/runtime/proc.go:290 +0x25

goroutine 3 [GC sweep wait]:
runtime.gopark(0x1338038, 0x18e3860, 0xc, 0x14, 0x1)
        /usr/local/src/go/1.19/go/src/runtime/proc.go:363 +0xfd fp=0xc00008f768 sp=0xc00008f738 pc=0x44317d
runtime.goparkunlock(0x0?, 0x0?, 0x0?, 0x0?)
        /usr/local/src/go/1.19/go/src/runtime/proc.go:369 +0x2a fp=0xc00008f798 sp=0xc00008f768 pc=0x44320a
runtime.bgsweep(0x0?)
        /usr/local/src/go/1.19/go/src/runtime/mgcsweep.go:278 +0x98 fp=0xc00008f7c8 sp=0xc00008f798 pc=0x42ab38
runtime.gcenable.func1()
        /usr/local/src/go/1.19/go/src/runtime/mgc.go:178 +0x26 fp=0xc00008f7e0 sp=0xc00008f7c8 pc=0x41ef86
runtime.goexit()
        /usr/local/src/go/1.19/go/src/runtime/asm_amd64.s:1594 +0x1 fp=0xc00008f7e8 sp=0xc00008f7e0 pc=0x4755c1
created by runtime.gcenable
        /usr/local/src/go/1.19/go/src/runtime/mgc.go:178 +0x6b

goroutine 4 [GC scavenge wait]:
runtime.gopark(0x1338038, 0x18e40c0, 0xd, 0x14, 0x2)
        /usr/local/src/go/1.19/go/src/runtime/proc.go:363 +0xfd fp=0xc00008ff48 sp=0xc00008ff18 pc=0x44317d
runtime.goparkunlock(0x1392098?, 0x1?, 0x0?, 0x0?)
        /usr/local/src/go/1.19/go/src/runtime/proc.go:369 +0x2a fp=0xc00008ff78 sp=0xc00008ff48 pc=0x44320a
runtime.(*scavengerState).park(0x18e40c0)
        /usr/local/src/go/1.19/go/src/runtime/mgcscavenge.go:389 +0x4b fp=0xc00008ffa0 sp=0xc00008ff78 pc=0x4287cb
runtime.bgscavenge(0x0?)
        /usr/local/src/go/1.19/go/src/runtime/mgcscavenge.go:617 +0x45 fp=0xc00008ffc8 sp=0xc00008ffa0 pc=0x428da5
runtime.gcenable.func2()
        /usr/local/src/go/1.19/go/src/runtime/mgc.go:179 +0x26 fp=0xc00008ffe0 sp=0xc00008ffc8 pc=0x41ef26
runtime.goexit()
        /usr/local/src/go/1.19/go/src/runtime/asm_amd64.s:1594 +0x1 fp=0xc00008ffe8 sp=0xc00008ffe0 pc=0x4755c1
created by runtime.gcenable
        /usr/local/src/go/1.19/go/src/runtime/mgc.go:179 +0xaa

goroutine 18 [finalizer wait]:
runtime.gopark(0x1338038, 0x1915bc0, 0x10, 0x14, 0x1)
        /usr/local/src/go/1.19/go/src/runtime/proc.go:363 +0xfd fp=0xc00008e5f8 sp=0xc00008e5c8 pc=0x44317d
runtime.goparkunlock(0x0?, 0x0?, 0x0?, 0xc00008e660?)
        /usr/local/src/go/1.19/go/src/runtime/proc.go:369 +0x2a fp=0xc00008e628 sp=0xc00008e5f8 pc=0x44320a
runtime.runfinq()
        /usr/local/src/go/1.19/go/src/runtime/mfinal.go:180 +0x107 fp=0xc00008e7e0 sp=0xc00008e628 pc=0x41e027
runtime.goexit()
        /usr/local/src/go/1.19/go/src/runtime/asm_amd64.s:1594 +0x1 fp=0xc00008e7e8 sp=0xc00008e7e0 pc=0x4755c1
created by runtime.createfing
        /usr/local/src/go/1.19/go/src/runtime/mfinal.go:157 +0x45

rax    0x0
rbx    0x7f4321f1da00
rcx    0x7f4320664ce1
rdx    0x0
rdi    0x2
rsi    0x7ffdbf324320
rbp    0x0
rsp    0x7ffdbf324320
r8     0x0
r9     0x7ffdbf324320
r10    0x8
r11    0x246
r12    0x7f4321e7e9a8
r13    0x7f4321cb77d8
r14    0x7f4321e55ad0
r15    0x7ffdbf3246b0
rip    0x7f4320664ce1
rflags 0x246
cs     0x33
fs     0x0
gs     0x0

Debugger finished with the exit code 0
```

When fs mode is local, process successfully start, but remote dns does not work and app return error: `no such host`.
For test mirrord with current config I created simple go server and try to start mirroring. It works correctly, without any errors.

It will be great to get a help to solve this problem. Thank you

### PR DESCRIPTION
**Pull request by aviramha**
add replace_with_fallback to use when functions might not be available in Go hook context [specifically, fstat]

Fix Go calling `fstat` syscall when it's not hooked (for some reason?)
Closes #2254

### CODE REVIEW

File: mirrord/layer/src/macros.rs [hunk 1]
```diff
@@ -60,6 +60,58 @@ macro_rules! replace {
...
+///   is created by `hook_fn`.
+/// - `$fallback_fn`: stores the libc function to fallback to if the hook fails.
+///
+/// ## Examples
+///
+/// - Replacing [`libc::close`] with [`close_detour`](crate::close_detour):
+///
+/// ```rust, no_run
+/// unsafe {
+///     replace!(&mut hook_manager, "close", close_detour, FnClose, FN_CLOSE);
```
Comment (meowjesty) [inline mirrord/layer/src/macros.rs:89]:
```suggestion
///     replace_with_fallback!(hook_manager, "close", close_detour, FnClose, FN_CLOSE, libc::close);
```

File: mirrord/layer/src/file/hooks.rs [hunk 2]
```diff
@@ -1212,7 +1212,14 @@ pub(crate) unsafe fn enable_file_hooks(hook_manager: &mut HookManager) {
             FN___LXSTAT64
         );
         replace!(hook_manager, "lstat", lstat_detour, FnLstat, FN_LSTAT);
-        replace!(hook_manager, "fstat", fstat_detour, FnFstat, FN_FSTAT);
+        crate::replace_with_fallback!(
```
Comment (meowjesty) [inline mirrord/layer/src/file/hooks.rs:1215]: Import it instead of `crate::`

**Review by meowjesty — APPROVED** [review]
:goberserk:

File: mirrord/layer/src/file/hooks.rs (hunk 2 above)
Comment (aviramha) [reply mirrord/layer/src/file/hooks.rs:1215]: for a one time?

Comment (meowjesty) [reply mirrord/layer/src/file/hooks.rs:1215]: It's fine either way, just that the others we import.

In the end, doesn't really matter, up to you.

**Pull request merged**